core/src/main/java/org/apache/calcite/rel/type/RelDataType.java

```diff
@@ -40,6 +40,11 @@
   int SCALE_NOT_SPECIFIED = Integer.MIN_VALUE;
   int PRECISION_NOT_SPECIFIED = -1;
 
+  /**
+   * Suffix for the digests of non-nullable types.
+   */
+  String NON_NULLABLE_TYPE_DIGEST_SUFFIX = " NOT NULL";
```

Comment:
Frankly speaking, I do not like how this leaks to the public API.
Could you declare the constant in a Calcite-private location to prevent it accidental use in the client code?

Reply:
Me too, i prefer the string constant `" NOT NULL"` directly because it is stragit-forward.

Reply:
I understand @vlsi 's concern. I guess it would be better to define this constant e.g. in RelDataTypeImpl rather than RelDataType.
@danny0405 I think it is cleaner to "centralize" this constant in one single place.

Reply:
@vlsi 's concern sounds reasonable, so I have moved the literal to class `RelDataTypeImpl` according to @rubenada 's suggestion.

IMO, extracting a constant to a static final field makes the code more structured and maintainable, and it also helps avoiding some bugs and typos (like the one in this issue).